# Guía práctica de Modelos de Regresión

Primera parte: https://youtu.be/9IZ6OPQWtpw

<h1 style="color:purple"> Métricas para Regresión </h1>

<h3 style="color:purple">Error absoluto máximo (M)</h3>

$y$ : valor verdadero 

$\hat{y}$ : valor predicho 

$$\text{M}(y, \hat{y}) = max(\left| y_i - \hat{y}_i \right|)$$


In [15]:
from sklearn.metrics import max_error

y_verdadero = [1, 2, 3, 4, 5]
y_predicho = [1, 2, 3, 4, -5]
max_error(y_verdadero, y_predicho)

10


<h3 style="color:purple">Error absoluto medio (mean absolute error - MAE)</h3>

$y$ : valor verdadero 

$\hat{y}$ : valor predicho 

$n$ : tamaño de la muestra 

$$\text{MAE}(y, \hat{y}) = \frac{1}{n} \sum_{i=1}^{n} \left| y_i - \hat{y}_i \right|$$




In [2]:
from sklearn.metrics import mean_absolute_error

y_verdadero = [1, 2, 3, 4, 5]
y_predicho = [1, 2, 3, 4, -5]
mean_absolute_error(y_verdadero, y_predicho)

2.0

<h3 style="color:purple">Error cuadrático medio (mean squared error - MSE)</h3>
$y$ : valor verdadero 

$\hat{y}$ : valor predicho 

$n$ : tamaño de la muestra 

$$\text{MSE}(y, \hat{y}) = \frac{1}{n} \sum_{i=1}^{n} (y_i - \hat{y}_i)^2$$


In [16]:
from sklearn.metrics import mean_squared_error

y_verdadero = [1, 2, 3, 4, 5]
y_predicho = [1, 2, 3, 4, -5]
mean_squared_error(y_verdadero, y_predicho)

20.0

<h3 style="color:purple">Suma de los cuadrados de los residuos (RSS)</h3>

$y$ : valor verdadero 

$\hat{y}$ : valor predicho 

$n$ : tamaño de la muestra 

$$\text{RSS}(y, \hat{y}) = \sum_{i=1}^{n} (y_i - \hat{y}_i)^2$$

In [4]:
from sklearn.metrics import mean_squared_error

y_verdadero = [1, 2, 3, 4, 5]
y_predicho = [1, 2, 3, 4, -5]
mean_squared_error(y_verdadero, y_predicho)*len(y_predicho)

100.0

<h3 style="color:purple">Raíz cuadrada del error cuadrático medio (RMSE)</h3>

$y$ : valor verdadero 

$\hat{y}$ : valor predicho 

$n$ : tamaño de la muestra 

$$\text{RMSE}(y, \hat{y}) = \sqrt{\frac{1}{n} \sum_{i=1}^{n} (y_i - \hat{y}_i)^2}$$

In [17]:
from sklearn.metrics import mean_squared_error

y_verdadero = [1, 2, 3, 4, 5]
y_predicho = [1, 2, 3, 4, -5]
mean_squared_error(y_verdadero, y_predicho, squared=False)

/opt/conda/lib/python3.12/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


4.47213595499958

<h3 style="color:purple">R^2 (Coeficiente de determinación)</h3>

$y$ : valor verdadero 

$\hat{y}$ : valor predicho 

$\bar{y}$ : promedio de los valores verdaderos

$n$ : tamaño de la muestra 

$$R^2(y, \hat{y}) = 1 - \frac{\sum_{i=1}^{n} (y_i - \hat{y}_i)^2}{\sum_{i=1}^{n} (y_i - \bar{y})^2}$$

In [18]:
from sklearn.metrics import r2_score

y_verdadero = [1, 2, 3, 4, 5]
y_predicho = [1, 2, 3, 4, 2]
r2_score(y_verdadero, y_predicho)

0.09999999999999998

Spiess, Andrej-Nikolai, and Natalie Neumeyer. "An evaluation of R2 as an inadequate measure for nonlinear models in pharmacological and biochemical research: a Monte Carlo approach." BMC pharmacology 10.1 (2010): 1-11.

Segunda parte

<h1 style="color:purple"> Caso práctico </h1>

In [31]:
from sklearn.linear_model import Ridge
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error, median_absolute_error, r2_score

import numpy as np
import pandas as pd
data = [
    {"edad": 27, "estudio": 9, "horas": 40, "valor_salario": 1296.33},
    {"edad": 30, "estudio": 9, "horas": 40, "valor_salario": 2017.40},
    {"edad": 42, "estudio": 9, "horas": 40, "valor_salario": 3905.04},
    {"edad": 68, "estudio": 9, "horas": 25, "valor_salario": 3356.25},
    {"edad": 34, "estudio": 8, "horas": 40, "valor_salario": 824.42},
    {"edad": 26, "estudio": 13, "horas": 50, "valor_salario": 676.35},
    {"edad": 23, "estudio":  4, "horas": 25, "valor_salario": 226.80},
    {"edad": 20, "estudio": 10, "horas": 40, "valor_salario": 1719.41},
    {"edad": 21, "estudio": 10, "horas": 15, "valor_salario": 225.37},
    {"edad": 41, "estudio": 10, "horas": 40, "valor_salario": 1346.16},    
]
# Separar en X e y
df_sample = pd.DataFrame(data)
X = df_sample.drop(columns=["valor_salario"])
y = df_sample["valor_salario"]
# X: features, y: target
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

ridge = Ridge(alpha=10)  # prueba con alpha=1, 10, 100
ridge.fit(X_train, y_train)
print("Coeficientes:", ridge.coef_)
print("Intercepto:", ridge.intercept_)

y_pred = ridge.predict(X_test)
def mape(y_true, y_pred, eps=1e-8): # 0.00000001
    y_true = np.asarray(y_true); y_pred = np.asarray(y_pred)
    denom = np.maximum(np.abs(y_true), eps)
    return np.mean(np.abs((y_true - y_pred) / denom)) * 100.0

print("MAE:", mean_absolute_error(y_test, y_pred))
print("MSE:", mean_squared_error(y_test, y_pred))
print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred)))
print("MAPE:", mape(y_test, y_pred))
print("R²:", r2_score(y_test, y_pred))

Coeficientes: [63.37192712  4.09170438 22.41432287]
Intercepto: -1434.4563869077951
MAE: 332.7081443450531
MSE: 191687.41536966097
RMSE: 437.8212139328803
MAPE: 25.97428656078818
R²: 0.7612391912649961


## 📏 Interpretación del error relativo (MAE y RMSE)

**Rango de salarios**
- Salario mínimo ≈ **225 soles**  
- Salario máximo ≈ **3905 soles**  
- **Rango** = 3905 – 225 = **3680 soles**

---

### 🔹 MAE (Error Absoluto Medio)
$ \text{MAE} = 332.71 $

Comparación con el rango:

$$
\frac{332.71}{3680} \approx 0.0904 \Rightarrow 9.0\%
$$

👉 En promedio, el modelo se equivoca en **~9% del rango de salarios**.


---

### 🔹 RMSE (Raíz del Error Cuadrático Medio)
$ \text{RMSE} = 437.82 $

Comparación con el rango:

$$
\frac{437.82}{3680} \approx 0.1189 \Rightarrow 11.9\%
$$

👉 El error típico (penalizando más los errores grandes) es de **~12% del rango**.

### ?
El error de este modelo Ridge se mueve **entre 9% y 12% del rango total de salarios**.  
Esto significa que, aunque el modelo explica bien la variación (R² ≈ 0.76), todavía hay un margen de error considerable, especialmente para valores bajos de salario.

---

### 🔹 MAPE (Error Porcentual Absoluto Medio)
$ \text{MAPE} = 25.97\% $

Interpretación:

👉 En promedio, el modelo se equivoca en **~26% respecto a cada valor real de salario**.  
Esto significa que, observación por observación, la predicción suele desviarse un 26% del valor correcto.  
⚠️ El MAPE puede verse más alto en registros donde el salario real es pequeño, porque el cálculo divide entre $y_{true}$.

---

### 🔹 R² (Coeficiente de Determinación)
$ R^2 = 0.76 $

Interpretación:

👉 El modelo logra explicar aproximadamente **el 76% de la variabilidad de los salarios**.  
El 24% restante corresponde a factores que el modelo lineal no captura o a ruido en los datos.  
Un $R^2$ cercano a 1 indica buen ajuste; en este caso, el resultado es **aceptable considerando solo 10 registros**.



## 📊 Uso e interpretación de R²

| Aspecto | Modelos Lineales (ej. LinearRegression, Ridge, Lasso) | Modelos No Lineales (ej. RandomForest, XGBoost, SVR, Redes Neuronales) |
|---------|--------------------------------------------------------|------------------------------------------------------------------------|
| **Definición** | Proporción de varianza de \(y\) explicada por la combinación lineal de las variables. | Proporción de varianza de \(y\) explicada por el modelo, sin asumir linealidad. |
| **Fórmula** | $R^2 = 1 - \dfrac{\sum_{i=1}^{n}(y_i - \hat{y}_i)^2}{\sum_{i=1}^{n}(y_i - \bar{y})^2}$ | Misma fórmula, aplicada a los residuos del modelo no lineal. |
| **Rango típico** | 0 a 1 (puede ser negativo si el modelo es peor que la media). | Igual: 0 a 1 (también puede ser negativo). |
| **Interpretación** | Indica cuánto de la variabilidad se explica mediante relaciones lineales entre X y y. | Indica la capacidad predictiva del modelo (sin relación con la forma funcional). |
| **Ventajas** | Fácil de interpretar en términos de “% de varianza explicada”. | Se puede usar como métrica comparativa entre modelos. |
| **Limitaciones** | Puede inflarse al añadir más variables, aunque no sean útiles (por eso se usa el \(R^2\) ajustado). | No tiene significado geométrico; solo mide qué tan bien predice el modelo frente a la media. |

---

### ✅ Conclusión
- **R² aplica en cualquier modelo de regresión (lineal o no lineal).**  
- En modelos lineales tiene un **significado estadístico fuerte**.  
- En modelos no lineales se interpreta solo como **métrica de desempeño predictivo**.



In [33]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, mean_absolute_error, median_absolute_error, r2_score
import numpy as np
import pandas as pd

# ===============================
# Dataset de ejemplo
# ===============================
data = [
    {"edad": 27, "estudio": 9, "horas": 40, "valor_salario": 1296.33},
    {"edad": 30, "estudio": 9, "horas": 40, "valor_salario": 2017.40},
    {"edad": 42, "estudio": 9, "horas": 40, "valor_salario": 3905.04},
    {"edad": 68, "estudio": 9, "horas": 25, "valor_salario": 3356.25},
    {"edad": 34, "estudio": 8, "horas": 40, "valor_salario": 824.42},
    {"edad": 26, "estudio": 13, "horas": 50, "valor_salario": 676.35},
    {"edad": 23, "estudio":  4, "horas": 25, "valor_salario": 226.80},
    {"edad": 20, "estudio": 10, "horas": 40, "valor_salario": 1719.41},
    {"edad": 21, "estudio": 10, "horas": 15, "valor_salario": 225.37},
    {"edad": 41, "estudio": 10, "horas": 40, "valor_salario": 1346.16},    
]

# Separar en X e y
df_sample = pd.DataFrame(data)
X = df_sample.drop(columns=["valor_salario"])
y = df_sample["valor_salario"]

# Train/test split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# ===============================
# Modelo: Random Forest Regressor
# ===============================
rfr = RandomForestRegressor(
    n_estimators=200,       # número de árboles
    max_depth=None,         # sin límite de profundidad
    random_state=42,
    n_jobs=-1               # usar todos los núcleos disponibles
)

rfr.fit(X_train, y_train)
y_pred = rfr.predict(X_test)

# ===============================
# Métricas
# ===============================
def mape(y_true, y_pred, eps=1e-8):
    y_true = np.asarray(y_true); y_pred = np.asarray(y_pred)
    denom = np.maximum(np.abs(y_true), eps)
    return np.mean(np.abs((y_true - y_pred) / denom)) * 100.0

print("MAE:", mean_absolute_error(y_test, y_pred))
print("MSE:", mean_squared_error(y_test, y_pred))
print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred)))
print("MAPE:", mape(y_test, y_pred))
print("R²:", r2_score(y_test, y_pred))
print("Importancia de variables:", rfr.feature_importances_)


MAE: 978.6090000000017
MSE: 973688.8069645633
RMSE: 986.7567111322645
MAPE: 266.3041412401466
R²: -0.21280119802729391
Importancia de variables: [0.84109226 0.08390124 0.0750065 ]


## 📊 Interpretación de resultados con RandomForestRegressor (RFR)

### 🔹 Métricas de error
- **MAE = 978.61**  
👉 En promedio, el modelo se equivoca en **~979 soles** al predecir los salarios.

- **RMSE = 986.76**  
👉 El error típico es de **~987 soles**. Comparado con salarios de 225 a 3905 soles, representa un desvío de entre **25% y 40%**.

- **MAPE = 266.30%**  
👉 El error porcentual es muy alto: las predicciones difieren en promedio un **266% respecto al valor real**.  
⚠️ Esto pasa porque hay salarios muy pequeños (ej. 225 soles) → cualquier error se convierte en un porcentaje enorme.

---

### 🔹 R² (Coeficiente de determinación)
- **R² = -0.21**  
👉 El modelo predice **peor que simplemente usar la media de los salarios**.  
Esto ocurre porque el dataset es muy pequeño (10 registros) y el test set solo tiene 2 observaciones → las métricas se vuelven inestables.

---

### 🔹 Importancia de variables
- **edad**: 84%  
- **estudio**: 8%  
- **horas**: 7%  

👉 El modelo da un peso muy alto a la **edad**, pero con tan pocos datos esta conclusión no es confiable.

---

### ✅ Conclusión
- El RFR necesita **más datos** para mostrar su verdadero potencial.  
- Con solo 10 registros, las métricas salen poco estables (MAPE inflado y R² negativo).  
- Los errores (≈1000 soles) son grandes frente al rango de salarios.  
- **Edad** aparece como variable principal, aunque esto puede ser solo efecto del dataset pequeño.


## 📊 Comparación: Ridge vs RandomForestRegressor (RFR)

| Métrica | Ridge | RFR |
|---------|-------|-----|
| **MAE** | 332.71 | 978.61 |
| **RMSE** | 437.82 | 986.76 |
| **MAPE** | 25.97 % | 266.30 % |
| **R²** | 0.76 | -0.21 |

---

### 🔹 Interpretación
- **Ridge**  
  👉 Se ajusta mejor a los datos: error medio bajo (~333 soles) y explica el **76% de la variabilidad** de los salarios.  
  👉 El error relativo (MAPE) es razonable (~26%).  

- **RandomForestRegressor (RFR)**  
  👉 Con tan pocos registros (10), no logra generalizar bien → el **R² es negativo** y el MAPE se dispara (>266%).  
  👉 El error absoluto (~979 soles) es mucho más grande en relación al rango de salarios.  
  👉 Señala a la **edad** como variable dominante, pero esto es poco confiable con tan poca data.  

---

### ✅ Conclusión
- En este dataset pequeño, **Ridge es claramente superior a RFR**.  
- El **RFR necesita muchos más datos** para mostrar su verdadero potencial y aprovechar su naturaleza no lineal.  
- Para enseñanza, este ejemplo es útil para mostrar cómo **el tamaño de la muestra impacta en el desempeño de distintos modelos**.
